src/rtc/mosaic_geobursts.py

```diff
@@ -186,73 +273,184 @@ def compute_weighted_mosaic_array(list_rtc_images, list_nlooks,
+            offset_imgy = int((list_geo_transform[i,3] - ymax_mosaic) / posting_y + 0.5)
 
-        # TODO: if geogrid_in is None, check reprojection
+        if verbose:
+            print(f'        image offset (x, y): ({offset_imgx}, {offset_imgy})')
 
-        # calculate the burst RTC's offset wrt. the output mosaic in the image coordinate
-        offset_imgx = int((list_geo_transform[i,0] - xmin_mosaic) / posting_x + 0.5)
-        offset_imgy = int((list_geo_transform[i,3] - ymax_mosaic) / posting_y + 0.5)
+        if path_nlooks is not None:
+            nlooks_gdal_ds = gdal.Open(path_nlooks, gdal.GA_ReadOnly)
+            arr_nlooks = nlooks_gdal_ds.ReadAsArray()
+            invalid_ind = np.isnan(arr_nlooks)
+            arr_nlooks[invalid_ind] = 0.0
+        else:
+            arr_nlooks = 1
 
-        if verbose:
-            print(f'image offset [x, y] = [{offset_imgx}, {offset_imgy}]')
-        raster_rtc = gdal.Open(path_rtc,0)
-        arr_rtc = raster_rtc.ReadAsArray()
+        rtc_image_gdal_ds = gdal.Open(path_rtc, gdal.GA_ReadOnly)
+
+        for i_band in range(num_bands):
 
-        #reshape arr_rtc when it is a singleband raster: to make it compatible in the for loop below
-        if num_bands==1:
-            arr_rtc=arr_rtc.reshape((1, arr_rtc.shape[0], arr_rtc.shape[1]))
+            band_ds = rtc_image_gdal_ds.GetRasterBand(i_band+1)
+            arr_rtc = band_ds.ReadAsArray()
+            if i_band == 0:
+                length, width = arr_rtc.shape
 
-        # Replace NaN values with 0
-        arr_rtc[np.isnan(arr_rtc)] = 0.0
+            if mosaic_mode.lower() == 'average':
+                # Replace NaN values with 0
+                arr_rtc[np.isnan(arr_rtc)] = 0.0
 
-        raster_nlooks = gdal.Open(path_nlooks, 0)
-        arr_nlooks = raster_nlooks.ReadAsArray()
-        invalid_ind = np.isnan(arr_nlooks) 
-        arr_nlooks[invalid_ind] = 0.0
+                arr_numerator[i_band,
+                            offset_imgy: offset_imgy + length,
+                            offset_imgx: offset_imgx + width] += \
+                    arr_rtc * arr_nlooks
 
-        for i_band in range(num_bands):
+                if path_nlooks is not None:
+                    arr_denominator[offset_imgy: offset_imgy + length,
+                                    offset_imgx: offset_imgx + width] += arr_nlooks
+                else:
+                    arr_denominator[offset_imgy: offset_imgy + length,
+                                    offset_imgx: offset_imgx + width] += np.asarray(
+                        arr_rtc > 0, dtype=np.byte)
+
+                continue
+
+            arr_temp = arr_numerator[i_band, offset_imgy: offset_imgy + length,
+                                     offset_imgx: offset_imgx + width].copy()
+
+            if i_band == 0 and mosaic_mode.lower() == 'first':
+                ind = np.where(np.isnan(arr_temp))
```

Comment:
```suggestion
                ind = np.isnan(arr_temp)
```
I think you can use [Boolean array indexing](https://numpy.org/doc/stable/user/basics.indexing.html#boolean-array-indexing) here

Reply:
These indexes are assigned in the first iteration and kept during the other iterations of the for loop. Using `np.where` would probably require less memory here.

Reply:
in case of boolean array: `n * m` bytes where `n` and `m` are the # of rows and columns of the array respectively:
in case of np.where: `2 * 8 * p` bytes where `p` is the locations to the pixel

Therefore if the area with `nan` value is greater than 1/16 of the total area, then boolean indexing would be more memory efficient. Otherwise, `np.where` would be better.

This is the information I can provide you to decide which one to use. It's your call to decide which indexing you use.

Reply:
It makes sense, @seongsujeong . Yes, we'll have more than 1/16 of the image with NaNs, so just accepted your suggestion!